pbsmetrics/pbsmetrics.go

```diff
@@ -0,0 +1,350 @@
+package pbsmetrics
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/rcrowley/go-metrics"
+)
+
+type Metrics struct {
+	metricsRegistry     metrics.Registry
+	RequestMeter        metrics.Meter
+	AppRequestMeter     metrics.Meter
+	NoCookieMeter       metrics.Meter
+	SafariRequestMeter  metrics.Meter
+	SafariNoCookieMeter metrics.Meter
+	ErrorMeter          metrics.Meter
+	InvalidMeter        metrics.Meter
+	RequestTimer        metrics.Timer
+	CookieSyncMeter     metrics.Meter
+	// Metrics for OpenRTB requests specifically. So we can track what % of RequestsMeter are OpenRTB
+	// and know when legacy requests have been abandoned.
+	ORTBRequestMeter metrics.Meter
+
+	AdapterMetrics map[openrtb_ext.BidderName]*AdapterMetrics
+
+	exchanges []openrtb_ext.BidderName
+}
+
+type AdapterMetrics struct {
+	NoCookieMeter     metrics.Meter
+	ErrorMeter        metrics.Meter
+	NoBidMeter        metrics.Meter
+	TimeoutMeter      metrics.Meter
+	RequestMeter      metrics.Meter
+	RequestTimer      metrics.Timer
+	PriceHistogram    metrics.Histogram
+	BidsReceivedMeter metrics.Meter
+}
+
+// Create a new Metrics object with all blank metrics object. This may also be useful for
+// testing routines to ensure that no metrics are written anywhere.
+
+func NewBlankMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName) *Metrics {
```

Comment:
I'm a little confused by this... I don't see this method used anywhere except in `_test` code.

The comment says you want to prevent metrics from being written... but I'm not sure why? I understand why you wouldn't want them _exported_ anywhere... but it's not the Meters which do that. It's [these exporters](https://github.com/prebid/prebid-server/blob/master/pbs_light.go#L796).

Even with Meters, Timers, etc which don't record any data, an Exporter can still write those 0's to an external source... so I don't think all this code helps in the way you think it does.

Reply:
I do use this in the following NewMetrics() function. The idea is we can turn on and off individual metrics from collection without refactoring the code that generates the metrics or writing our own exporter that does not assume we want to export everything in the registry.

The double New...() is that the NewBlankMetrics() insures every metrics hook is filled with a blank (no-op) metrics handler, and then the NewMetrics() fills in real metrics handlers for the metrics we currently care about. So we never have to refactor any code if we want to temporarily suspend an old metric. The blank metrics will slightly improve performance in test code as well, but that by itself is not worth this effort.

One use case is having one set of metrics we care about for day-to-day monitoring in production, and then a superset of this for debugging we can switch on when we suspect a problem. This allows us to have a slim set of standard metrics to optimize performance and metrics storage space, yet be able to generate a very rich set of metrics when we have an issue we wish to debug.

Reply:
ok.

In all honesty... if I had to take bets, I'd wager that these are more likely to do harm than good.

You're right that this would make debug-level monitoring a bit easier to implement. On the other hand, if we forget to override the blank metric when we add a new one, we'll need to debug why we're not getting some expected metric in influx.

That said: I don't think it's a huge deal either way. All the code is contained within this module, and none of the public APIs are affected (at least, none of the public APIs which are being used). So I'm happy to let it ride and see what happens.

The second reviewer might want to chime in here too though, for a third opinion.